File: src/2022/day09/src/main/java/ca/terrylockett/aoc2022/day09/Day09Runner.java
package ca.terrylockett.aoc2022.day09;

import ca.terrylockett.aoccommon.resources.Resources;

public class Day09Runner {

	public static void main(String[] args) throws Exception {
		String inputFile = Resources.getInputFilePath("input.txt").orElseThrow();

		Rope ropePart1 = new Rope(2);
		ropePart1.processMoves(inputFile);

		System.out.println("Day 09 Part 01: " + ropePart1.getTailKnot().getVisitedLocationsCount());

		Rope ropePart2 = new Rope(10);
		ropePart2.processMoves(inputFile);

		System.out.println("Day 09 Part 02: " + ropePart2.getTailKnot().getVisitedLocationsCount());
	}

}

File: src/2023/day09/src/main/kotlin/ca/terrylockett/aoc2023/day09/Day09Runner.kt
package ca.terrylockett.aoc2023.day09

import ca.terrylockett.aoccommon.resources.Resources
import java.io.File

typealias Report = List<Int>

fun main() {
	val inputFile: String = Resources.getInputFilePath("input.txt").orElseThrow()
	println("2023 day09 part1: ${part1(inputFile)}")
	println("2023 day09 part2: ${part2(inputFile)}")
}

fun part1(inputFilePath: String): Int {
	val reports = createReports(inputFilePath)
	
	var total = 0
	for (report in reports) {
		val nextToken = report.last() + calcNextElement(report)
		total += nextToken
	}

	return total
}

fun part2(inputFilePath: String): Int {
	val reports = createReports(inputFilePath)
	
	var total = 0
	for (report in reports) {
		val nextToken = report.first() - calcPreviousElement(report)
		total += nextToken
	}

	return total
}

fun createReports(inputFilePath: String): ArrayList<Report> {
	val reports = ArrayList<Report>()

	File(inputFilePath).forEachLine { line ->
		val report: Report = line.split(" ").map(Integer::parseInt).toList()
		reports.add(report)
	}

	return reports
}

fun calcNextElement(report: Report): Int {
	val diffList = ArrayList<Int>()

	for (i in 0..report.size - 2) {
		val diff = report[i + 1] - report[i]
		diffList.add(diff)
	}

	if (diffList.stream().allMatch { it == 0 }) {
		return diffList.last()
	}

	return diffList.last() + calcNextElement(diffList)
}

fun calcPreviousElement(report: Report): Int {
	val diffList = ArrayList<Int>()

	for (i in 0..report.size - 2) {
		val diff = report[i + 1] - report[i]
		diffList.add(diff)
	}

	if (diffList.stream().allMatch { it == 0 }) {
		return diffList.first()
	}

	return diffList.first() - calcPreviousElement(diffList)
}
